lib/spack/spack/reporters/extract.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import os
+import re
+import xml.sax.saxutils
+from datetime import datetime
+
+import llnl.util.tty as tty
+
+log_regexp = re.compile(r"^==> \[([0-9:.\-]*)(?:, [0-9]*)?\] (.*)")
+returns_regexp = re.compile(r"\[([0-9 ,]*)\]")
+
+skip_msgs = ["Testing package", "Results for"]
+skip_regexps = [re.compile(r"{0}".format(msg)) for msg in skip_msgs]
+status_values = ["FAILED", "PASSED", "NO-TESTS"]
+status_regexps = [re.compile(r"^({0})".format(stat)) for stat in status_values]
+
+
+def add_part_output(part, line):
+    if part:
+        part['loglines'].append(xml.sax.saxutils.escape(line))
+
+
+def elapsed(current, previous):
+    if not (current and previous):
+        return 0
+
+    diff = current - previous
+    tty.debug('elapsed = %s - %s = %s' % (current, previous, diff))
+    return diff.total_seconds()
+
+
+def expected_failure(line):
+    if not line:
+        return False
+
+    match = returns_regexp.search(line)
+    xfail = '0' not in match.group(0) if match else False
+    return xfail
+
+
+def new_part():
+    return {
+        'command': None,
+        'completed': None,
+        'desc': None,
+        'elapsed': None,
+        'name': None,
+        'loglines': [],
+        'output': None,
+        'status': None,
+    }
+
+
+def part_name(source):
+    # TODO: Should be passed the package prefix and only remove it
+    elements = []
+    for e in source.replace("'", "").split(' '):
+        elements.append(os.path.basename(e) if os.sep in e else e)
+    return '_'.join(elements)
+
+
+def process_part_end(part, curr_time, last_time):
+    if part:
+        if not part['elapsed']:
+            part['elapsed'] = elapsed(curr_time, last_time)
+
+        stat = part['status']
+        part['completed'] = 'Completed' if stat else 'Unknown'
+        part['output'] = '\n'.join(part['loglines'])
+
+
+def timestamp(time_string):
+    return datetime.strptime(time_string, '%Y-%m-%d-%H:%M:%S.%f')
+
+
+def skip(line):
+    for regex in skip_regexps:
+        match = regex.search(line)
+        return match
+
+
+def status(desc, line):
+    for regex in status_regexps:
+        match = regex.search(line)
+        if match:
+            stat = match.group(0)
+            xfail = stat == 'PASSED' and expected_failure(desc)
+            return 'xfail' if xfail else stat.lower()
```

Comment:
CDash is currently pretty rigid in its test statuses being either "passed", "failed", or "notrun". Instead of introducing a new status type here, I recommend that you mark the test as "failed" and provide additional details in the `completed` field. Here's an example of what this would look like in the list of tests on CDash:

<img width="1080" alt="Screen Shot 2022-01-28 at 3 21 42 PM" src="https://user-images.githubusercontent.com/224720/151616197-5f3ce0f1-3c0c-44aa-86a5-c07fdd82e6d5.png">

Reply:
Since we want to know if a package does not have tests to run, I created a dummy test part so we can flag/track such packages.  What do others think of reporting packages in this manner? 

An example output from running at the command line for the `blt` package is given below.

```
<?xml version="1.0" encoding="UTF-8"?>
<Site BuildName="blt-m6hnxcd346nyhmzwagcmyfzv35piomjy"
      BuildStamp="20220128-1706-Experimental"
      Name="ruby961"
      Generator="spack-0.17.1-1009-ec03cebdcd"
      Hostname="ruby961"
      OSName="Linux"
      OSRelease="3.10.0-1160.45.1.1chaos.ch6.x86_64"
      VendorString="cascadelake"
>
  <Testing>
    <StartTime>1643418411</StartTime>
    <Test Status="notrun">
      <Name>blt</Name>
      <FullCommand>Not Applicable</FullCommand>
      <Results>
        <NamedMeasurement type="numeric/double" name="Execution Time">
          <Value>0.0</Value>
        </NamedMeasurement>
        <NamedMeasurement type="text/string" name="Completion Status">
          <Value>No tests to run</Value>
        </NamedMeasurement>
      </Results>
    </Test>
    <EndTestTime>1643418411</EndTestTime>
  </Testing>
```

Reply:
The changes to address this issue are reflected in the description's updated outputs.